composer/utils/checkpoint.py

```diff
@@ -601,6 +602,57 @@ def load_checkpoint(
     return rng_state_dicts
 
 
+def dist_cp_load(
+    state_dict: dict[str, Any],
+    storage_reader: StorageReader,
+    load_planner: Optional[LoadPlanner] = None,
+):
+    if version.parse(torch.__version__) >= version.parse('2.4.0'):
+        from torch.distributed.checkpoint.utils import CheckpointException
+        try:
+            dist_cp.load(
+                state_dict=state_dict,
+                storage_reader=storage_reader,
+                planner=load_planner,
+            )
+        except CheckpointException as e:
+            checkpoint_metadata = storage_reader.read_metadata().state_dict_metadata
+            if 'state.metadata' in checkpoint_metadata and 'state.metadata.composer_env_info.composer_version' not in checkpoint_metadata:
+                # Torch 2.4 changed the way how state dict is flattened. It broke backward compatibility.
+                # Torch issue: https://github.com/pytorch/pytorch/issues/133923.
+                # We override the traverse_state_dict so that the load planner could
+                # use the old way of flattening the state dict
```

Comment:
@mvpatel2000 , i put the comment here where we do the monkey patch. I didn't add doc string to the function since this function include the normal working path.